File: Data Structure & Algorithm/Template/Binary Search Tree/Binary Search Tree.ts
// Binary Search Tree

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class BinarySearchTree {
  root: TreeNode | null;

  constructor(node: TreeNode | null) {
    this.root = node;
  }

  searchNode(value: number) {
    const recursive = (node: TreeNode | null): boolean => {
      if (!node) {
        return false;
      }

      if (value < node.val) {
        return recursive(node.left);
      } else if (value > node.val) {
        return recursive(node.right);
      } else {
        return true;
      }
    };

    return recursive(this.root);
  }

  insertNode(new_value: number) {
    if (!this.root) {
      this.root = new TreeNode(new_value);
      return true;
    }

    const insert = (node: TreeNode): boolean => {
      if (new_value < node.val) {
        if (!node.left) {
          node.left = new TreeNode(new_value);
          return true;
        }
        return insert(node.left);
      } else if (new_value > node.val) {
        if (!node.right) {
          node.right = new TreeNode(new_value);
          return true;
        }
        return insert(node.right);
      } else {
        return false;
      }
    };

    return insert(this.root);
  }

  deleteNode(value: number) {
    const findRightMin = (node: TreeNode, parent: TreeNode) => {
      let current_node = node;
      let current_parent = parent;
      while (current_node.left) {
        current_parent = current_node;
        current_node = current_node.left;
      }
      return [current_node, current_parent];
    };

    const recursive = (node: TreeNode | null, parent: TreeNode | null) => {
      if (!node) {
        return;
      }

      if (value < node.val) {
        recursive(node.left, node);
      } else if (value > node.val) {
        recursive(node.right, node);
      } else {
        if (!node.left && !node.right) {
          parent!.left = null;
          parent!.right = null;
        } else if (!node.left || !node.right) {
          parent!.val = node.val;
          parent!.left = null;
          parent!.right = null;
        } else {
          const [min_node, min_node_parent] = findRightMin(node.right, node);
          node.val = min_node.val;
          if (min_node_parent.left === min_node) {
            min_node_parent.left = min_node.right;
          } else {
            min_node_parent.right = min_node.right;
          }
        }
      }
    };

    recursive(this.root, null);
  }

  preorder() {
    const res: number[] = [];

    const recursive = (node: TreeNode | null) => {
      if (!node) {
        return;
      }

      res.push(node.val);
      recursive(node.left);
      recursive(node.right);
    };

    recursive(this.root);
    return res;
  }

  inorder() {
    const res: number[] = [];

    const recursive = (node: TreeNode | null) => {
      if (!node) {
        return;
      }

      recursive(node.left);
      res.push(node.val);
      recursive(node.right);
    };

    recursive(this.root);
    return res;
  }

  postorder() {
    const res: number[] = [];

    const recursive = (node: TreeNode | null) => {
      if (!node) {
        return;
      }

      recursive(node.left);
      recursive(node.right);
      res.push(node.val);
    };

    recursive(this.root);
    return res;
  }
}

// test
const tree = new BinarySearchTree(null);
tree.insertNode(8);
tree.insertNode(14);
tree.insertNode(10);
tree.insertNode(13);
tree.insertNode(3);
tree.insertNode(1);
tree.insertNode(6);
tree.insertNode(4);
tree.insertNode(7);
const insert_res1 = tree.preorder();
console.log(insert_res1);
const search_res1 = tree.searchNode(14);
console.log(search_res1);
const search_res2 = tree.searchNode(5);
console.log(search_res2);

tree.deleteNode(8);
const delete_res1 = tree.preorder();
console.log(delete_res1);

tree.deleteNode(3);
const delete_res2 = tree.preorder();
console.log(delete_res2);

File: Other/Review/Phase 5/Review1/binary-search-tree/binary-search-tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type BinarySearchTree struct {
	root *TreeNode
}

// NewTreeNode is a constructor for TreeNode
func NewTreeNode(val int) *TreeNode {
	return &TreeNode{Val: val, Left: nil, Right: nil}
}

// NewBinarySearchTree is a constructor for BinarySearchTree
func NewBinarySearchTree(node *TreeNode) *BinarySearchTree {
	return &BinarySearchTree{root: node}
}

func (bst *BinarySearchTree) SearchNode(value int) bool {
	var recursive func(node *TreeNode) bool
	recursive = func(node *TreeNode) bool {
		if node == nil {
			return false
		}

		if value < node.Val {
			return recursive(node.Left)
		} else if value > node.Val {
			return recursive(node.Right)
		} else {
			return true
		}
	}

	return recursive(bst.root)
}

func (bst *BinarySearchTree) InsertNode(newValue int) bool {

	if bst.root == nil {
		bst.root = NewTreeNode(newValue)
		return true
	}

	var recursive func(node *TreeNode) bool
	recursive = func(node *TreeNode) bool {

		if newValue < node.Val {
			if node.Left == nil {
				node.Left = NewTreeNode(newValue)
				return true
			}
			return recursive(node.Left)
		} else if newValue > node.Val {
			if node.Right == nil {
				node.Right = NewTreeNode(newValue)
				return true
			}
			return recursive(node.Right)
		} else {
			return false
		}
	}

	return recursive(bst.root)
}

func (bst *BinarySearchTree) DeleteNode(value int) bool {
	var recursive func(node *TreeNode, parent *TreeNode) bool
	recursive = func(node *TreeNode, parent *TreeNode) bool {
		if node == nil {
			return false
		}

		if value < node.Val {
			return recursive(node.Left, node)
		} else if value > node.Val {
			return recursive(node.Right, node)
		} else {
			if node.Left == nil || node.Right == nil {
				var child *TreeNode
				if node.Left != nil {
					child = node.Left
				} else {
					child = node.Right
				}

				if parent.Left == nil {
					bst.root = child
				} else if parent.Left == node {
					parent.Left = child
				} else {
					parent.Right = child
				}
			} else {
				currentMinNode := node.Right
				parentNode := node

				for currentMinNode.Left != nil {
					parentNode = currentMinNode
					currentMinNode = currentMinNode.Left
				}

				node.Val = currentMinNode.Val

				if parentNode.Left == currentMinNode {
					parentNode.Left = currentMinNode.Right
				} else {
					parentNode.Right = currentMinNode.Right
				}
			}

			return true
		}
	}

	return recursive(bst.root, nil)
}

func (bst *BinarySearchTree) Preorder() []int {
	res := make([]int, 0)

	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node == nil {
			return
		}

		res = append(res, node.Val)
		recursive(node.Left)
		recursive(node.Right)
	}

	recursive(bst.root)
	return res
}

func (bst *BinarySearchTree) Inorder() []int {
	res := make([]int, 0)

	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node == nil {
			return
		}

		recursive(node.Left)
		res = append(res, node.Val)
		recursive(node.Right)
	}

	recursive(bst.root)
	return res
}

func (bst *BinarySearchTree) Postorder() []int {
	res := make([]int, 0)

	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node == nil {
			return
		}

		recursive(node.Left)
		recursive(node.Right)
		res = append(res, node.Val)
	}

	recursive(bst.root)
	return res
}

// test
func main() {
	tree := NewBinarySearchTree(nil)
	tree.InsertNode(8)
	tree.InsertNode(14)
	tree.InsertNode(10)
	tree.InsertNode(13)
	tree.InsertNode(3)
	tree.InsertNode(1)
	tree.InsertNode(6)
	tree.InsertNode(4)
	tree.InsertNode(7)
	insert_res1 := tree.Preorder()
	fmt.Println(insert_res1)
	search_res1 := tree.SearchNode(14)
	fmt.Println(search_res1)
	search_res2 := tree.SearchNode(5)
	fmt.Println(search_res2)

	tree.DeleteNode(8)
	delete_res1 := tree.Preorder()
	fmt.Println(delete_res1)

	tree.DeleteNode(3)
	delete_res2 := tree.Preorder()
	fmt.Println(delete_res2)
}
